File: client/ts/Prism.ts
import type { udp } from "bun";
import DataPacket from "./proto/DataPacket";
import { PacketType } from "./proto/Types";
import { createDecipheriv, createCipheriv, randomBytes } from "crypto";

const IV_LENGTH = 12;
const TAG_LENGTH = 16;

export default class Prism {
  private pendingAcks: Map<number, [() => void, () => void]>;
  private socket!: udp.Socket<"buffer">;

  private host: string;
  private port: number;

  private key: Buffer;

  private closed: boolean;

  public constructor(host: string, port: number, key: string) {
    this.pendingAcks = new Map();
    this.host = host;
    this.port = port;
    this.key = Buffer.from(key);
    this.closed = true;
  }

  public async start() {
    this.closed = false;
    this.socket = await Bun.udpSocket({
      connect: {
        port: this.port,
        hostname: this.host,
      },
      socket: {
        data: this.onData.bind(this),
      },
    });
  }

  private onData(
    sock: udp.Socket<"buffer">,
    data: Buffer<ArrayBufferLike>
  ): void | Promise<void> {
    const raw = Buffer.from(data);
    if (raw.length < IV_LENGTH + TAG_LENGTH + 9) {
      if (raw[0] == PacketType.ACK) {
        return this.handleAckPacket(raw);
      }
    }

    const iv = raw.subarray(0, IV_LENGTH); 
    const ciphertext = raw.subarray(IV_LENGTH, raw.length - TAG_LENGTH);
    const tag = raw.subarray(raw.length - TAG_LENGTH);

    if (iv.length !== IV_LENGTH || tag.length !== TAG_LENGTH) {
      console.error("Invalid IV or tag length");
      return;
    }

    const dec = this.decrypt(iv, ciphertext, tag);
    if (!dec) {
      console.error("Failed to decrypt data packet");
      return;
    }

    if (dec[0] == PacketType.DATA) {
      return this.handleDataPacket(dec, sock);
    }
  }

  public async send(data: Uint8Array, requestType: number): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const { iv, ciphertext, tag } = this.encrypt(Buffer.from(data));

      const seq = Math.floor(Math.random() * 0xffffffff);
      const header = Buffer.alloc(9);
      header.writeUInt8(requestType, 0);
      header.writeUInt32BE(seq, 1);
      const combinedBuffer = Buffer.concat([header, data]);
      let checksum = Bun.hash.crc32(combinedBuffer);
      header.writeUInt32BE(checksum, 5);

      const packet = Buffer.concat([header, iv, ciphertext, tag]);
      this.pendingAcks.set(seq, [resolve, reject]);
      // @ts-expect-error
      this.socket.send(packet);

      setTimeout(() => {
        if (this.pendingAcks.has(seq)) {
          this.pendingAcks.delete(seq);
          if (!this.closed) {
            console.error(`ACK not received for seqNum: ${seq}`);
          } else {
            console.error("Socket closed, not sending reject");
          }
        }
      }, 5000);
    });
  }

  private encrypt(data: Uint8Array): {
    iv: Buffer;
    ciphertext: Buffer;
    tag: Buffer;
  } {
    const iv = randomBytes(12);
    const cipher = createCipheriv("aes-256-gcm", this.key, iv);
    const ciphertext = Buffer.concat([cipher.update(data), cipher.final()]);
    const tag = cipher.getAuthTag();
    return { iv, ciphertext, tag };
  }

  private decrypt(iv: Buffer, ciphertext: Buffer, tag: Buffer): Buffer | null {
    try {
      const decipher = createDecipheriv("aes-256-gcm", this.key, iv);
      decipher.setAuthTag(tag);
      const decrypted = Buffer.concat([
        decipher.update(ciphertext),
        decipher.final(),
      ]);
      return decrypted;
    } catch (error) {
      console.log(error);
      return null;
    }
  }

  private handleDataPacket(data: Buffer, sock: udp.Socket<"buffer">) {
    const seq = data.readUInt32BE(1);
    const checksum = data.readUInt32BE(5);
    
    const packet = DataPacket.decode(Buffer.from(data));
    console.log(
      `Received Data Packet (seqNum: ${seq}, checksum: ${checksum}, data: ${JSON.stringify(
        packet.toJSON(),
        null,
        4
      )})`
    );

    const ack = Buffer.alloc(5);
    ack.writeUInt8(PacketType.ACK, 0);
    ack.writeUInt32BE(seq, 1);
    // @ts-expect-error
    sock.send(ack);

    switch (packet.type) {
      case 3:
        this.handleAuthResponse(packet);
        break;
    }
  }

  private handleAckPacket(raw: Buffer) {
    const seq = raw.readUInt32BE(1);

    if (this.pendingAcks.has(seq)) {
      const [resolve] = this.pendingAcks.get(seq)!;
      resolve();
      this.pendingAcks.delete(seq);
      console.log(`ACK confirmed, removed seqNum ${seq} from pending list`);
    } else {
      console.warn(`Received ACK for unknown seqNum: ${seq}`);
    }
  }

  private handleAuthResponse(packet: DataPacket) {
    const authResponse = packet.authResponse!;
    if (authResponse.status === 1) {
      console.log("Auth failed: ", authResponse.message);
      this.socket.close();
    }
  }

  public getHost(): string {
    return this.host;
  }

  public getPort(): number {
    return this.port;
  }
}

File: client/go/prism.go
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"hash/crc32"
	"math/rand"
	"net"
	"sync"
	"time"

	"google.golang.org/protobuf/proto"
)

const IV_LENGTH = 12
const TAG_LENGTH = 16

const (
	DATA = iota + 1
	ACK
	NACK
)

type Prism struct {
	host        string
	port        int
	conn        net.Conn
	pendingAcks sync.Map

	sec *security
}

func NewPrism(host string, port int, key string) *Prism {
	return &Prism{
		host: host,
		port: port,
		sec:  newSecurity(key),
	}
}

func (p *Prism) Start() error {
	conn, err := net.Dial("udp", "127.0.0.1:6969")
	if err != nil {
		return err
	}
	p.conn = conn

	go p.listen()
	return nil
}

func (p *Prism) listen() {
	buffer := make([]byte, 1024)
	for {
		n, err := p.conn.Read(buffer)
		if err != nil {
			fmt.Println("Error reading from UDP:", err)
			continue
		}

		if n < IV_LENGTH+TAG_LENGTH+9 {
			if buffer[0] == ACK {
				seq := binary.BigEndian.Uint32(buffer[1:5])
				p.handleAckPacket(seq)
			}
			continue
		}

		iv := buffer[:IV_LENGTH]
		ciphertext := buffer[IV_LENGTH : n-TAG_LENGTH]
		tag := buffer[n-TAG_LENGTH : n]

		if len(iv) != IV_LENGTH || len(tag) != TAG_LENGTH {
			fmt.Println("Invalid IV or tag length")
			continue
		}

		dec, err := p.sec.decrypt(iv, ciphertext, tag)
		if err != nil {
			fmt.Println("Failed to decrypt data packet", err)
			continue
		}

		typeID := dec[0]
		seq := binary.BigEndian.Uint32(dec[1:5])

		switch typeID {
		case 1:
			p.handleDataPacket(dec, seq)
		default:
			fmt.Println("Unknown packet type received")
		}
	}
}

func (p *Prism) handleDataPacket(data []byte, seq uint32) {
	seq = binary.BigEndian.Uint32(data[1:5])
	checksum := binary.BigEndian.Uint32(data[5:9])

	var pk DataPacket
	if err := proto.Unmarshal(data[9:], &pk); err != nil {
		fmt.Println("Failed to unmarshal data packet:", err)
		return
	}

	if crc32.ChecksumIEEE(data[9:]) != checksum {
		fmt.Println("Checksum mismatch, dropping packet")
		return
	}

	switch pk.Type {
	case 3:
		fmt.Println("Received AuthResponse for seq:", seq)
		ar := pk.GetAuthResponse()
		fmt.Println("AuthResponse:", ar)
	default:
		fmt.Println("Unknown packet type received")
	}

	ack := make([]byte, 5)
	ack[0] = 2
	binary.BigEndian.PutUint32(ack[1:], seq)
	p.conn.Write(ack)
}

func (p *Prism) handleAckPacket(seq uint32) {
	if ch, ok := p.pendingAcks.Load(seq); ok {
		close(ch.(chan struct{}))
		p.pendingAcks.Delete(seq)
		fmt.Printf("ACK confirmed, removed seq %d from pending list\n", seq)
	} else {
		fmt.Printf("Received ACK for unknown seq: %d\n", seq)
	}
}

func (p *Prism) Send(data []byte, requestType int) error {
	iv, ciphertext, tag := p.sec.encrypt(data)
	seq := rand.Uint32()
	header := make([]byte, 9)
	header[0] = byte(requestType)
	binary.BigEndian.PutUint32(header[1:], seq)

	combined := append(header, data...)
	checksum := crc32.ChecksumIEEE(combined)
	binary.BigEndian.PutUint32(header[5:], checksum)

	packet := append(header, iv...)
	packet = append(packet, ciphertext...)
	packet = append(packet, tag...)

	ackChan := make(chan struct{})
	p.pendingAcks.Store(seq, ackChan)

	if _, err := p.conn.Write(packet); err != nil {
		p.pendingAcks.Delete(seq)
		return err
	}

	select {
	case <-ackChan:
		return nil
	case <-time.After(5 * time.Second):
		p.pendingAcks.Delete(seq)
		return errors.New(fmt.Sprintf("ACK not received for seq: %d", seq))
	}
}

const (
	ADDR = "127.0.0.1"
	PORT = 6969
	KEY  = "secret-auth-key-123============="
)

func main() {
	prism := NewPrism(ADDR, PORT, KEY)
	if err := prism.Start(); err != nil {
		fmt.Println("Error starting UDP client:", err)
		return
	}
	login := Login{
		Service: "default_service",
		Token:   "default_token",
	}

	rawLogin, err := proto.Marshal(&login)
	if err != nil {
		panic(err)
	}

	err = prism.Send(rawLogin, DATA)
	if err != nil {
		fmt.Println("Failed to send login:", err)
	}

	data := DataPacket{
		Type: 4,
		Payload: &DataPacket_Update{
			Update: &Update{
				Name:         "test-name",
				Value:        "10",
				Type:         "KILLS",
				PersistCache: true,
			},
		},
	}

	rawData, err := proto.Marshal(&data)
	if err != nil {
		panic(err)
	}

	err = prism.Send(rawData, DATA)
	if err != nil {
		fmt.Println("Failed to send data packet:", err)
	}

	select {}
}
